Support for Metal Gear Solid 4

cipherxof opened the pull request:
MGS4 is fairly annoying to work with because each stage is actually a completely different executable, and it turns out each executable has the camera code duplicated. So, naturally the offsets are shifted every time you enter into a new area. I was able to find the pointers for all of the camera data in the base executable, which points to the offsets in whatever the currently loaded stage is.

As for why the 3PP uses shorts instead of floats, I'm not sure. Maybe they are Halfs or something.

Enjoy :D

**Known Issues**
- The mouse does not work on the Rex stage.
- You cannot move the camera horizontally when aiming with the knife in third person.

isJuhn commented on KAMI.Core/Games/MetalGearSolid4.cs:
```diff
@@ -0,0 +1,65 @@
+{
+    public class MetalGearSolid4 : Game<HAVACamera>
+    {
+        uint m_base_address;
+        uint m_camera_address;
```
Maybe there's better names for these, since both of them seem to be some kind of "camera base addresses"

isJuhn reviewed (commented):
Looks good, just some nitpicks.

cipherxof commented on KAMI.Core/Games/MetalGearSolid4.cs:
```diff
@@ -0,0 +1,65 @@
+            IPCUtils.WriteFloat(m_ipc, pCameraFPV + 0x4, m_camera.Hor);
+
+            // 3pp
+            uint pCameraTPP = IPCUtils.ReadU32(m_ipc, m_base_address + 0x4);
```
I could yes, but the pointer itself is at the base address as defined in Ghidra.

cipherxof commented on KAMI.Core/Games/MetalGearSolid4.cs:
```diff
@@ -0,0 +1,65 @@
+            ushort hor = IPCUtils.ReadU16(m_ipc, pCameraTPP);
+            ushort vert = IPCUtils.ReadU16(m_ipc, pCameraTPP + 0x8);
+            hor = (ushort)(hor + (-diffX * SensModifier) * 7000);
+            vert = (ushort)(vert + (diffY * SensModifier) * 7000);
```
Not really, I just used what felt right. I'm not really sure why shorts are used here honestly.

isJuhn replied:
Drakengard 3 uses something similar with signed 32 bit integers where a full rotation is 65536 units. My general philosophy for things like these is that a full rotation should always be the same no matter the game. That way your sensitivity in Kami is "universal", ofc there's different feels to different games and some are first person and others are third person (there's some ideas in my head about both of these, but that's for the future).

My suggestion is that you look at the resolution of a full rotation horizontally, and divide that by a full rotation of the camera you're using, in this case the HAVACamera is in degrees which means divide by 360.

isJuhn replied:
A dream reminded me that the HAVACamera uses radians, silly me

cipherxof replied:
Should be good to go now.

cipherxof commented:
Updated.

- Fixed a crash that would happen in some stages when changing the camera interpolation.
- Fixed some issues that would cause the camera to stop working at random stages / areas.

isJuhn commented:
Thanks :)
